### PR TITLE
CNV-18089: OVN-Kubernetes secondary networks

### PR DESCRIPTION
**Pull request by sjhala-ccs**
Version(s): 4.14+

Issue: [CNV-18089](https://issues.redhat.com//browse/CNV-18089)

Link to docs preview: https://64435--docspreview.netlify.app/openshift-enterprise/latest/virt/vm_networking/virt-connecting-vm-to-ovn-secondary-network

QE review:
- [x] QE has approved this change.

### CODE REVIEW

**Comment by maiqueb** [comment]
/cc

Thanks for this PR !

File: modules/virt-creating-layer2-nad-cli.adoc
Comment (maiqueb) [inline modules/virt-creating-layer2-nad-cli.adoc:36]: nit: it can actually be 0.4.0 as well.

Gist is, it **cannot** be >= 1.0.0

File: modules/virt-creating-layer2-nad-cli.adoc
Comment (maiqueb) [inline modules/virt-creating-layer2-nad-cli.adoc:27]: the user can also configure the MTU of the network. 

Should that (optional) knob be referenced in the example ? 

Keep in mind it must take into account the geneve tunnel overhead.

File: modules/virt-creating-layer2-nad-cli.adoc
Comment (maiqueb) [inline modules/virt-creating-layer2-nad-cli.adoc:37]: maybe indicate this is useful to interconnect VMs in different namespaces ? 

I'm aware that information is already there - but implicit. 

Do we want to be more explicit ?

File: modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc
Comment (maiqueb) [inline modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc:57]: maybe we should use a newer image here ? This one is ooooooold.

Maybe use the same one we're using elsewhere ?

File: modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc
Comment (maiqueb) [inline modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc:49]: I wonder if we want to have 2 interfaces in the example - or just one. 

What do we do on the examples for other secondary networks types ?

**Review by maiqueb — COMMENTED** [review]
Initial (very ...) shallow review. I only took a look into the provided YAMLs.

File: modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc
Comment (sjhala-ccs) [reply modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc:49]: SR-IOV example: https://docs.openshift.com/container-platform/4.14/virt/vm_networking/virt-connecting-vm-to-sriov.html#virt-attaching-vm-to-sriov-network_virt-connecting-vm-to-sriov

Linux bridge example: https://docs.openshift.com/container-platform/4.14/virt/vm_networking/virt-connecting-vm-to-linux-bridge.html#virt-attaching-vm-secondary-network-cli_virt-connecting-vm-to-linux-bridge

File: modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc
Comment (sjhala-ccs) [reply modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc:57]: We actually do not include the `volumes` stanza in the SR-IOV or Linux bridge documentation. What are the most relevant sections of the `VirtualMachine` manifest that we need to include in this example?

File: modules/virt-creating-layer2-nad-cli.adoc
Comment (sjhala-ccs) [reply modules/virt-creating-layer2-nad-cli.adoc:27]: I have added the `mtu` parameter and description. Let me know if that is what you meant by your comment.

File: modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc
Comment (maiqueb) [reply modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc:49]: Right ! Let's follow their lead then.

Thank you.

File: modules/virt-creating-layer2-nad-cli.adoc
Comment (maiqueb) [reply modules/virt-creating-layer2-nad-cli.adoc:27]: Hm, the default is not 1300 !!! 

It's chosen by the kernel - but I don't think it is 1300. 

Can we just leave it out in the open w/ `The default value is automatically set by the kernel.` 

set / chosen ?

File: modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc
Comment (maiqueb) [reply modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc:57]: I had assumed this would be a copy / paste able example. Looking at the other examples - [bridge](https://docs.openshift.com/container-platform/4.14/virt/vm_networking/virt-connecting-vm-to-linux-bridge.html#virt-attaching-vm-secondary-network-cli_virt-connecting-vm-to-linux-bridge) / [sriov](https://docs.openshift.com/container-platform/4.14/virt/vm_networking/virt-connecting-vm-to-sriov.html#virt-attaching-vm-to-sriov-network_virt-connecting-vm-to-sriov) - it looks it isn't.

I would argue the `cloudinitdisk` is a must since that is how the user would configure the IP on the interface - which would be borderline useless without it... 

But the same can be said about the other examples (none are showing cloud init configuration), and the name of the article / heading is `Connecting a VM to an XYZ network` . It says nothing about L3. Thus, I would keep it simple and do something like:

```yaml
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachine
metadata:
  name: vm-server
spec:
  template:
    spec:
      domain:
        devices:
          interfaces:
          - name: default
            masquerade: {}
          - name: flatl2-overlay (1)
            bridge: {}
      networks:
      - name: default
        pod: {}
      - name: flatl2-overlay  (2)
        multus:
          networkName: l2-network (3)
```

I would add procedures pointing out the name of the interface **must** match the name of the network, as done in [bridge](https://docs.openshift.com/container-platform/4.14/virt/vm_networking/virt-connecting-vm-to-linux-bridge.html#virt-attaching-vm-secondary-network-cli_virt-connecting-vm-to-linux-bridge) / [sriov](https://docs.openshift.com/container-platform/4.14/virt/vm_networking/virt-connecting-vm-to-sriov.html#virt-attaching-vm-to-sriov-network_virt-connecting-vm-to-sriov).

Comment (sjhala-ccs) [reply modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc:57]: Done

Comment (maiqueb) [reply modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc:57]: FWIW, I think you forgot to push the new changes. I still see the "old" code - with the img.

Comment (maiqueb) [inline modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc:57]: FWIW, I think you forgot to push the new changes. I still see the "old" code - with the img.

Comment (sjhala-ccs) [reply modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc:57]: Yes, pushing the updates now

File: modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc
Comment (maiqueb) [inline modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc:39]: sorry for having missed this before. Maybe - for transparency - we should also ensure the namespace appears here ?... 

This is especially important given openshift's `project` notion. 

I've checked the other examples at it seems we only list the namespace on [bridge](https://docs.openshift.com/container-platform/4.14/virt/vm_networking/virt-connecting-vm-to-linux-bridge.html#virt-attaching-vm-secondary-network-cli_virt-connecting-vm-to-linux-bridge); for [sriov](https://docs.openshift.com/container-platform/4.14/virt/vm_networking/virt-connecting-vm-to-sriov.html#virt-attaching-vm-to-sriov-network_virt-connecting-vm-to-sriov), we do not.

I think it should be aligned across all 3 . Either they all feature the namespace, or none does. Same for the procedure description.

@phoracek or for some reason, for sriov this doesn't allow specifying the namespace ? 

Should I open a doc bug to align the existing resources ?

**Review by maiqueb — COMMENTED** [review]
Other than this, I'm happy.

Let's just decide how we want the docs to be and I'll LGTM the PR.

Thank you for describing this feature in such a simple way. You made this look easy :)

File: modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc
Comment (phoracek) [reply modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc:39]: I would prefer to keep our example without the namespace for all. With that, people can copy paste the example and apply it directly to whatever their currently selected namespace is.

To preserve the cross-namespace example, I would suggest we explain that in a callout.

I'll leave it to @sjhala-ccs to say whether we can handle it here or should create a BZ for a follow-up PR.

Comment (maiqueb) [reply modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc:39]: As long as they're aligned, I'm happy. FWIW, you justification for removing is IMO solid.

File: modules/virt-creating-layer2-nad-cli.adoc
Comment (phoracek) [inline modules/virt-creating-layer2-nad-cli.adoc:37]: Sounds like a bug, I would not bother calling it out here. We don't expect people to change this, they should treat it as a magic number. "This value must be less than `1.0.0`." assumes users know what CNI versions are available. Or just say that the required value is "0.3.1".

File: modules/virt-creating-layer2-nad-cli.adoc
Comment (phoracek) [inline modules/virt-creating-layer2-nad-cli.adoc:49]: We use the actual name in the rest of our docs: https://docs.openshift.com/container-platform/4.14/virt/vm_networking/virt-connecting-vm-to-linux-bridge.html#virt-creating-linux-bridge-nad-cli_virt-connecting-vm-to-linux-bridge

File: modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc
Comment (phoracek) [inline modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc:32]:
```suggestion
          - name: secondary <1>
```

IMO the used topology is not relevant here and should be abstracted away from the VM

**Review by phoracek — CHANGES_REQUESTED** [review]

File: modules/virt-creating-layer2-nad-cli.adoc
Comment (sjhala-ccs) [reply modules/virt-creating-layer2-nad-cli.adoc:49]: It is not consistently the case in the rest of our docs. For example: https://docs.openshift.com/container-platform/4.14/virt/vm_networking/virt-using-dpdk-with-sriov.html#virt-configuring-vm-dpdk_virt-using-dpdk-with-sriov
Because the file name is a user-supplied value, I would prefer to keep this as is.

File: modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc
Comment (sjhala-ccs) [reply modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc:32]: This means that the `spec.networks.name` value should also be changed to `secondary`, correct?

File: modules/virt-creating-layer2-nad-cli.adoc
Comment (sjhala-ccs) [reply modules/virt-creating-layer2-nad-cli.adoc:37]: That is what it said before ([see comment](https://github.com/openshift/openshift-docs/pull/64435#discussion_r1321582610)) :-) Will change it back again.

Comment (maiqueb) [reply modules/virt-creating-layer2-nad-cli.adoc:37]: I'm OK with it FWIW, but again ... it can be seen as a bug, or just "we don't support conflist based configurations on this plugin at all".

I guess it's semantics. 

BTW. Why 0.3.1 and not 0.4.0 for instance ?...

File: modules/virt-creating-layer2-nad-cli.adoc
Comment (maiqueb) [reply modules/virt-creating-layer2-nad-cli.adoc:38]: I'm unsure. This is the procedure list. If you don't pay attention to the steps in something called "procedure list",  what are you going to pay attention to ? 

Not that I object.

File: modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc
Comment (maiqueb) [reply modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc:32]: yes

File: modules/virt-creating-layer2-nad-cli.adoc
Comment (sjhala-ccs) [reply modules/virt-creating-layer2-nad-cli.adoc:38]: Also, Asciidoctor doesn't support putting an admonition in the middle of a call out list, only when the code block is not part of an ordered list. In our scenario, it interrupts the call out list numbering.

File: modules/virt-creating-layer2-nad-cli.adoc
Comment (sjhala-ccs) [reply modules/virt-creating-layer2-nad-cli.adoc:37]: So, what's the verdict on this one? Should I keep or remove the second sentence in the call out description? Just say `The CNI specification version.`?

File: modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc
Comment (sjhala-ccs) [reply modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc:39]: I would suggest creating a GitHub issue or BZ to remove the namespace reference from the [bridge](https://docs.openshift.com/container-platform/4.14/virt/vm_networking/virt-connecting-vm-to-linux-bridge.html#virt-attaching-vm-secondary-network-cli_virt-connecting-vm-to-linux-bridge) example.

File: modules/virt-creating-layer2-nad-cli.adoc
Comment (phoracek) [reply modules/virt-creating-layer2-nad-cli.adoc:53]: Gotcha, all good then

File: modules/virt-creating-layer2-nad-cli.adoc
Comment (phoracek) [inline modules/virt-creating-layer2-nad-cli.adoc:30]:
```suggestion
            "name": "my-namespace-l2-network", <2>
```

File: modules/virt-creating-layer2-nad-cli.adoc
Comment (phoracek) [reply modules/virt-creating-layer2-nad-cli.adoc:38]: Understood. What about adding the namespace prefix to the example?

File: modules/virt-creating-layer2-nad-cli.adoc
Comment (phoracek) [inline modules/virt-creating-layer2-nad-cli.adoc:38]:
```suggestion
<2> The name of the network. This attribute is not namespaced. For example, you can have a network named `l2-network` referenced from two different `NetworkAttachmentDefinition` objects that exist in two different namespaces. This feature is useful to connect VMs in different namespaces.
```

File: modules/virt-creating-layer2-nad-cli.adoc
Comment (maiqueb) [reply modules/virt-creating-layer2-nad-cli.adoc]: Use the 0.3.1, as @phoracek suggests. 

Seems it's [the documented value for pods](https://docs.openshift.com/container-platform/4.13/networking/multiple_networks/configuring-additional-network.html#configuration-ovnk-network-plugin-json-object_configuring-additional-network).

File: modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc
Comment (maiqueb) [reply modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc:39]: Opened this github issue @sjhala-ccs .

Thanks.

File: modules/virt-creating-layer2-nad-cli.adoc
Comment (sjhala-ccs) [reply modules/virt-creating-layer2-nad-cli.adoc]: Sounds good

**Review by phoracek — APPROVED** [review]
Thanks!

File: modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc
Comment (stevsmit) [inline modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc:50]:
```suggestion
$ oc apply -f <filename>.yaml
```

File: virt/vm_networking/virt-connecting-vm-to-ovn-secondary-network.adoc
Comment (stevsmit) [inline virt/vm_networking/virt-connecting-vm-to-ovn-secondary-network.adoc:40]: Title of redirect page is "Configuration for an OVN-Kubernetes additional network" if you want them parallel.

**Comment by stevsmit** [comment]
@sjhala-ccs Great job. A few comments.

File: virt/vm_networking/virt-connecting-vm-to-ovn-secondary-network.adoc
Comment (sjhala-ccs) [reply virt/vm_networking/virt-connecting-vm-to-ovn-secondary-network.adoc:11]: Changing it to "by" which seems to fit better in this context

File: modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc
Comment (Anatw) [inline modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc:16]: The bellow yaml can't be created:
1. You must add running strategy under spec, or else you'll get this error:
The request is invalid: spec.running: One of Running or RunStrategy must be specified
2. Once created the VM will not start without a memory definition under spec.template.spec.domain.resources.requests.memory


Here is my suggestion for the yaml:
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachine
metadata:
  name: vm-server
spec:
  running: true
  template:
    spec:
      domain:
        devices:
          interfaces:
          - name: default
            masquerade: {}
          - name: secondary 
            bridge: {}
        resources:
          requests:
            memory: 1024Mi
      networks:
      - name: default
        pod: {}
      - name: secondary  
        multus:
          networkName: l2-network

File: modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc
Comment (maiqueb) [inline modules/virt-attaching-vm-to-ovn-secondary-nw-cli.adoc:20]: alpha3 is deprecated IIRC. 

We should use v1, like in this [example](https://docs.openshift.com/container-platform/4.13/virt/virtual_machines/virt-create-vms.html#virt-creating-vm-cli_virt-create-vms).

File: modules/virt-creating-layer2-nad-cli.adoc
Comment (maiqueb) [reply modules/virt-creating-layer2-nad-cli.adoc]: I am now re-reading this and I do wonder: how can the DHCP server deployed by the customer on the **physical** network impact the overlay ? (which is disconnected ...).

Am I missing something @phoracek ?

**Review by maiqueb — APPROVED** [review]
Thank you.

**Pull request merged**
